### PR TITLE
data-request-reward-collateral-ratio: initial draft

### PR DESCRIPTION
**Pull request by drcpu-github**
The protocol-enforced ratio still needs to be decided for the final WIP, I think both proposals are defensible.

I will also need to update the WIP-0023 link before both are merged in the main repository.

The reference implementation needs a lot of refinement.

### CODE REVIEW

File: data-request-reward-collateral-ratio.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+A potential downside of introducing a reward to collateral ratio at the node level is that it is possible to manipulate the reputation system by creating and solving data requests which other nodes will not solve. This proposal solves this issue by introducing this ratio at the protocol level. A data request which does not adhere to the minimum reward to collateral ratio will be considered as invalid. If a block is proposed including such a data request, it is also considered as invalid.
+
+One additional advantage of a protocol-enforced reward to collateral ratio is that several (griefing) attacks becoming significantly less interesting from an economic viewpoint. Without a reward to collateral ratio, tt's possible to create data requests that lock up collateral from other nodes or aim to introduce collateral slashing for virtually no money. For example, you could create malicious requests requiring 100 WIT collateral and set all rewards to 1 nanoWIT resulting in the data request being virtually free. If a reward to collateral ratio is introduced these type of griefing attacks will cost a tangible amount of money to execute.
```
Comment (aesedepece) [inline data-request-reward-collateral-ratio.md:29]:
```suggestion
One additional advantage of a protocol-enforced reward to collateral ratio is that several (griefing) attacks becoming significantly less interesting from an economic viewpoint. Without a reward to collateral ratio, it is possible to create data requests that lock up collateral from other nodes or aim to introduce collateral slashing for virtually no money. For example, you could create malicious requests requiring 100 WIT collateral and set all rewards to 1 nanoWIT resulting in the data request being virtually free. If a reward to collateral ratio is introduced these type of griefing attacks will cost a tangible amount of money to execute.
```

File: data-request-reward-collateral-ratio.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+## Specification
+
+### Introduction of a protocol-enforced data request reward collateral ratio
+
+Properly enforcing a reward to collateral ratio at the protocol level requires adding this ratio as a consensus constant. Adding a consensus constant will change the magic number nodes use to validate inter-node communication. Therefore, proper precautions need to be implemented such that this number only changes when the WIP is accepted.
+
+The moment the WIP is accepted, nodes which receive a data request that does not adhere to the reward to collateral ratio should refuse to add it to their local mempool.
```
Comment (aesedepece) [inline data-request-reward-collateral-ratio.md:37]: From a TAPI perspective, it's clearer to talk about _activation_ vs. _acceptance_.
```suggestion
The moment the WIP is activated, nodes which receive a data request that does not adhere to the reward to collateral ratio MUST refuse to add it to their local mempool.
```

File: data-request-reward-collateral-ratio.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+```
```
Comment (aesedepece) [inline data-request-reward-collateral-ratio.md:39]:
```suggestion
```rust
```

File: data-request-reward-collateral-ratio.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+    ...
+
+    self.dr_transactions.insert(key, (priority, dr_tx));
+}
+```
+
+Each and every block a node receives is validated which implies that all transactions in it are checked against a set of rules. If a node receives a block containing a data request that does not adhere to the required reward to collateral ratio, they should mark the block as invalid and refuse it as a candidate.
+
+```
```
Comment (aesedepece) [inline data-request-reward-collateral-ratio.md:56]:
```suggestion
```rust
```

File: data-request-reward-collateral-ratio.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    ...
+
+}
+```
+
+### Introduction of a configurable data request reward collateral ratio
+
+Next to a protocol-enforced reward to collateral ratio, node operators should be able to set a ratio they consider acceptable. This ratio cannot be set lower than the protocol-enforced ratio. Setting the ratio higher than the protocol-enforced ratio implies that these node operators will not add received data requests with a lower ratio to their local mempool. However, if they receive a block with a ratio which is lower than the ratio they accept but higher than the protocol-enforced ratio, they have to accept it as a valid candidate.
```
Comment (aesedepece) [inline data-request-reward-collateral-ratio.md:79]:
```suggestion
Next to a protocol-enforced reward to collateral ratio, node operators SHOULD be able to set a ratio they consider acceptable. This ratio MUST NOT be set lower than the protocol-enforced ratio. Setting the ratio higher than the protocol-enforced ratio implies that these node operators MAY refuse to add received data requests with a lower ratio to their local mempool. However, if they receive a block with a ratio which is lower than the ratio they accept but higher than the protocol-enforced ratio, they MUST accept it as a valid candidate.
```

**Review by aesedepece — CHANGES_REQUESTED** [review]
Some formatting comments are in.

**Comment by aesedepece** [comment]
In accordance with WIP-0001, acting as the WIP Editor, I hereby assign WIP number 0022 as proposed by @drcpu-github under the title _Data request reward collateral ratio_.

File: data-request-reward-collateral-ratio.md [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+
+One additional advantage of a protocol-enforced reward to collateral ratio is that several (griefing) attacks becoming significantly less interesting from an economic viewpoint. Without a reward to collateral ratio, it is possible to create data requests that lock up collateral from other nodes or aim to introduce collateral slashing for virtually no money. For example, you could create malicious requests requiring 100 WIT collateral and set all rewards to 1 nanoWIT resulting in the data request being virtually free. If a reward to collateral ratio is introduced these type of griefing attacks will cost a tangible amount of money to execute.
+
+## Specification
+
+### Introduction of a protocol-enforced data request reward collateral ratio
+
+Properly enforcing a reward to collateral ratio at the protocol level requires adding this ratio as a consensus constant. Adding a consensus constant will change the magic number nodes use to validate inter-node communication. Therefore, proper precautions need to be implemented such that this number only changes when the WIP is accepted.
```
Comment (aesedepece) [inline data-request-reward-collateral-ratio.md:35]:
```suggestion
Properly enforcing a reward to collateral ratio at the protocol level requires adding this ratio as a consensus constant. Adding a consensus constant will change the magic number nodes used to validate communication between nodes. Therefore, proper precautions need to be implemented such that this number only changes when the WIP is accepted.
```

File: data-request-reward-collateral-ratio.md [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+    let weight = f64::from(dr_tx.weight());
+    let priority = OrderedFloat(fee as f64 / weight);
+
+    ...
+
+    self.dr_transactions.insert(key, (priority, dr_tx));
+}
+```
+
+Each and every block a node receives is validated which implies that all transactions in it are checked against a set of rules. If a node receives a block containing a data request that does not adhere to the required reward to collateral ratio, they should mark the block as invalid and refuse it as a candidate.
```
Comment (aesedepece) [inline data-request-reward-collateral-ratio.md:54]:
```suggestion
Each and every block that a node receives is validated, which implies that all transactions in it are checked against a set of rules. If a node receives a block containing a data request that does not adhere to the required reward to collateral ratio, they should mark the block as invalid and refuse it as a candidate.
```

File: data-request-reward-collateral-ratio.md [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+    }
+
+    ...
+
+}
+```
+
+### Introduction of a configurable data request reward collateral ratio
+
+Next to a protocol-enforced reward to collateral ratio, node operators SHOULD be able to set a ratio they consider acceptable. This ratio MUST NOT be set lower than the protocol-enforced ratio. Setting the ratio higher than the protocol-enforced ratio implies that these node operators MAY refuse to add received data requests with a lower ratio to their local mempool. However, if they receive a block with a ratio which is lower than the ratio they accept but higher than the protocol-enforced ratio, they MUST to accept it as a valid candidate.
```
Comment (aesedepece) [inline data-request-reward-collateral-ratio.md:79]:
```suggestion
Next to a protocol-enforced reward to collateral ratio, node operators SHOULD be able to set a ratio they consider acceptable. This ratio MUST NOT be set lower than the protocol-enforced ratio. Setting the ratio higher than the protocol-enforced ratio implies that these node operators MAY refuse to add received data requests with a lower ratio to their local mempool. However, if they receive a block with a ratio which is lower than the ratio they accept but higher than the protocol-enforced ratio, they MUST accept it as a valid candidate.
```

File: data-request-reward-collateral-ratio.md [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+### Reward collateral ratio proposal
+
+There are two sensible values for a protocol-enforced reward to collateral ratio.
+
+The first one is that the ratio should offset the involuntary lie ratio. Assuming this is 0.2%, the minimum reward to collateral ratio should be 1 to 500.
+
+The second one can be derived from the maximum amount of witnesses in a data request which, following [WIP-0007](https://github.com/witnet/WIPs/blob/master/wip-0007.md), is 125. Assuming creating data requests requiring 125 witnesses has as goal to achieve maximum security, the reward for them should be proportional. This implies that such data request should pay out a reward equal to the required collateral or that the reward to collateral ratio should be 1 to 125.
```
Comment (aesedepece) [inline data-request-reward-collateral-ratio.md:87]:
```suggestion
The second one can be derived from the maximum amount of witnesses in a data request which, following [WIP-0007](https://github.com/witnet/WIPs/blob/master/wip-0007.md), is 125. Assuming creating data requests requiring 125 witnesses has as goal to achieve maximum security, the reward for them should be proportional. This implies that such data request should pay out a reward equal to the required collateral, or that the reward to collateral ratio should be 1 to 125.
```

**Review by aesedepece — CHANGES_REQUESTED** [review]

**Comment by tmpolaczyk** [comment]
> The second one can be derived from the maximum amount of witnesses in a data request which, following WIP-0007, is 125. Assuming creating data requests requiring 125 witnesses has as goal to achieve maximum security, the reward for them should be proportional. This implies that such data request should pay out a reward equal to the required collateral, or that the reward to collateral ratio should be 1 to 125.

Does that mean that the total cost of creating the data request would be equal to the collateral, in that case? Because I don't understand that requirement.

Overall I like the idea of the proposal, because it makes sense that increasing the required collateral should increase the cost of creating the data request. I guess the main blocker is deciding on the correct ratio?

And reminder that such a change will probably break most of the data requests deployed in ethereum, requiring a redeployment with higher reward fees. The good think is that users can start redeploying before this WIP is accepted, so there is plenty of time.

**Comment by drcpu-github** [comment]
> Does that mean that the total cost of creating the data request would be equal to the collateral, in that case? Because I don't understand that requirement.

Yes, that is exactly what it means. If the ratio is set at 1:125 and you create a data request with 125 witnesses, you'll pay the same total reward as the collateral you request from a single node.

> Overall I like the idea of the proposal, because it makes sense that increasing the required collateral should increase the cost of creating the data request. I guess the main blocker is deciding on the correct ratio?

I am indeed not quite sure what a decent ratio would be. I want the ratio to be high enough to make sense, but it should not inhibit adoption of Witnet by making requests too expensive.

> And reminder that such a change will probably break most of the data requests deployed in ethereum, requiring a redeployment with higher reward fees. The good think is that users can start redeploying before this WIP is accepted, so there is plenty of time.

Yes, if the ratio is set higher than 1:500, it will break all currently-deployed price feed request. I don't think there are many other recurring data requests deployed.

**Review by aesedepece — APPROVED** [review]
This proposal meets now all the WIP-0001 requirements to be published as a draft.

**Pull request closed without merging**